### PR TITLE
feat: add certificate issuance and bulk issuance functionality

### PR DESCRIPTION
**Pull request by ussyalfaks**
### 🔐 **Core Registration System**  
- **University Registration**  
  - Secure registration system for accredited universities  
  - Validates essential fields (name, address, country)  
- **Owner Authorization**  
  - Only contract owner can register universities  
  - Prevents unauthorized registrations  
- **Event Emissions**  
  - Registration triggers events for blockchain traceability  

### 📜 **Certificate Management**  
- **Single Certificate Issuance**  
  - Universities can issue certificates with:  
    - Metadata (encrypted)  
    - Hashed key (for verification)  
    - Unique certificate ID  
- **Bulk Certificate Issuance**  
  - Efficient batch processing for multiple certificates in one transaction  
- **University Validation**  
  - Certificate issuance restricted to registered universities only  
- **Secure Storage**  
  - Certificates stored using **deterministic hash-based key derivation**  

### 🔍 **Data Retrieval**  
- **Certificate Lookup**  
  - Retrieve certificates by ID using optimized hash function  
- **University Lookup**  
  - Query university details by wallet address  
- **Owner Verification**  
  - Helper function to verify contract interactions  

### 🛡️ **Security & Optimization**  
- **Hash Function**  
  - Robust string hashing for certificate IDs (`31x` multiplier method)  
- **Array Length Validation**  
  - Prevents data corruption in bulk operations  
- **Loop Optimization**  
  - Improved Cairo efficiency: `i != len` instead of `i < len`  
- **Ownership Checks**  
  - All sensitive operations validate caller permissions  

### 🧪 **Testing**  
- Comprehensive test suite covering:  
  - ✅ Regular and edge case scenarios  
  - ✅ Event emission verification  
  - ✅ Authorization checks  
  - ✅ Bulk operations  
  - ✅ Error handling  

## ⚙️ **Technical Details**  
- **Built on:** StarkNet (Cairo 1.0)  
- **Standards:** ERC-compatible design  
- **Storage:** Optimized for efficient lookups & minimal state changes

### CODE REVIEW

File: src/certiva.cairo [hunk 1]
```diff
@@ -27,10 +31,59 @@ pub mod Certiva {
...
     #[event]
     #[derive(Drop, starknet::Event)]
     pub enum Event {
         university_created: University,
+        certificate_issued: Certificate,
+        certificates_bulk_issued: BulkCertificatesIssued,
+    }
+
+    // Helper function to hash a ByteArray to a felt252
+    fn hash_certificate_id(certificate_id: ByteArray) -> felt252 {
```
Comment (JayWebtech) [inline src/certiva.cairo:59]: No need of this function. hashing will be done in FE

**Review by JayWebtech — COMMENTED** [review]
Also, create the interface in the /interfaces folder

**Comment by JayWebtech** [comment]
Thanks @ussyalfaks for your contribution
Certiva appreciates your efforts in helping to solve certificate fraud.
